File: python/src/aoc_2024_09.py
"""Advent of Code 2024 - Day 9."""

from dataclasses import dataclass
from sys import stdin


@dataclass
class Chunk:
    """One chunk on disk: file or free space."""

    id_: int | None  # File id (or None for free-space)
    blocks: int  # Number of blocks in the file
    index: int  # Index of first block of the file


def read_disk(s: str) -> list[Chunk]:
    """Read disk string into a list."""
    acc: list[Chunk] = []
    file_index: int = 0
    for i, c in enumerate(s):
        blocks = int(c)
        if i % 2 == 0:
            id_: int | None = i // 2
        else:
            id_ = None
        acc.append(Chunk(id_=id_, blocks=blocks, index=file_index))
        file_index += blocks
    return acc


def compact(disk: list[Chunk]) -> int:
    """Compact the disk and return the checksum."""
    left_chunk: int = 0  # Chunk we are currently summing
    right_chunk: int = len(disk) - 1  # Chunk we are transferring from
    assert disk[right_chunk].id_ is not None  # Check last chunk is a file

    total_file_blocks = sum(d.blocks for d in disk if d.id_ is not None)

    current_block: int = 0
    check_sum: int = 0
    left_summed: int = 0  # Blocks already check-summed from left_chunk
    right_transferred: int = 0  # Blocks transferred from right_chunk

    while current_block < total_file_blocks:
        assert disk[left_chunk].blocks != 0
        assert disk[right_chunk].blocks != 0
        match disk[left_chunk].id_:
            case None:
                right_file_id = disk[right_chunk].id_
                assert right_file_id is not None
                check_sum += right_file_id * current_block
                right_transferred += 1
                if right_transferred >= disk[right_chunk].blocks:
                    right_chunk -= 2
                    right_transferred = 0
            case file_id:
                check_sum += file_id * current_block
        left_summed += 1
        while left_summed >= disk[left_chunk].blocks:
            left_chunk += 1
            left_summed = 0
        current_block += 1
    return check_sum


if __name__ == "__main__":
    disk = read_disk(stdin.read().strip())
    print(compact(disk))

File: rust/src/bin/aoc_2024_09.rs
// Advent of Code 2024 - Day 9.

use std::io::{Read, stdin};

struct Chunk {
    id: Option<usize>,
    blocks: usize,
}

fn read_disk() -> Vec<Chunk> {
    let mut v: Vec<Chunk> = Vec::new();
    let mut buf = String::new();
    stdin().read_to_string(&mut buf).unwrap();
    for (i, ch) in buf.trim().chars().enumerate() {
        let blocks: usize = ch.to_digit(10).unwrap().try_into().unwrap();
        let id = if i % 2 == 0 { Some(i / 2) } else { None };
        v.push(Chunk { id, blocks });
    }
    v
}

fn compact(disk: &[Chunk]) -> usize {
    let total_file_blocks: usize = disk
        .iter()
        .filter(|chunk| chunk.id.is_some())
        .map(|chunk| chunk.blocks)
        .sum();

    let mut left_chunk: usize = 0;
    let mut right_chunk: usize = disk.len() - 1;

    let mut current_block: usize = 0;
    let mut check_sum: usize = 0;
    let mut left_summed: usize = 0;
    let mut right_transferred: usize = 0;

    assert!(disk[right_chunk].id.is_some());

    while current_block < total_file_blocks {
        assert!(disk[left_chunk].blocks > 0);
        assert!(disk[right_chunk].blocks > 0);
        if let Some(left_file_id) = disk[left_chunk].id {
            check_sum += left_file_id * current_block;
        } else {
            let right_file_id = disk[right_chunk].id.unwrap();
            check_sum += right_file_id * current_block;
            right_transferred += 1;
            if right_transferred >= disk[right_chunk].blocks {
                right_chunk -= 2;
                right_transferred = 0;
            }
        }
        left_summed += 1;
        while left_summed >= disk[left_chunk].blocks {
            left_chunk += 1;
            left_summed = 0;
        }
        current_block += 1;
    }
    check_sum
}

fn find_right_most_fit(
    current_chunk: usize,
    space: usize,
    disk: &[Chunk],
) -> Option<(usize, usize, usize)> {
    for i in (current_chunk..disk.len()).rev() {
        if let Some(i_id) = disk[i].id
            && disk[i].blocks <= space
        {
            return Some((i, i_id, disk[i].blocks));
        }
    }
    None
}

fn compact_whole(disk: &mut [Chunk]) -> usize {
    let mut current_chunk: usize = 0;
    let mut current_block: usize = 0;
    let mut check_sum: usize = 0;

    while current_chunk < disk.len() {
        let mut blocks = disk[current_chunk].blocks;
        if let Some(file_id) = disk[current_chunk].id {
            for _ in 0..blocks {
                check_sum += file_id * current_block;
                current_block += 1;
            }
        } else {
            while blocks > 0 {
                if let Some((r, r_id, r_blocks)) = find_right_most_fit(current_chunk, blocks, disk)
                {
                    for _ in 0..r_blocks {
                        check_sum += r_id * current_block;
                        current_block += 1;
                        blocks -= 1;
                    }
                    disk[r].id = None;
                } else {
                    current_block += blocks;
                    break;
                }
            }
        }
        current_chunk += 1;
    }
    check_sum
}

fn main() {
    let mut disk = read_disk();
    println!("{}", compact(&disk));
    println!("{}", compact_whole(&mut disk));
}
